osu.Game.Rulesets.Taiko/Difficulty/Skills/Colour.cs

```diff
@@ -49,95 +29,8 @@ public Colour(Mod[] mods)
 
         protected override double StrainValueOf(DifficultyHitObject current)
         {
-            // changing from/to a drum roll or a swell does not constitute a colour change.
-            // hits spaced more than a second apart are also exempt from colour strain.
-            if (!(current.LastObject is Hit && current.BaseObject is Hit && current.DeltaTime < 1000))
-            {
-                monoHistory.Clear();
-
-                var currentHit = current.BaseObject as Hit;
-                currentMonoLength = currentHit != null ? 1 : 0;
-                previousHitType = currentHit?.Type;
-
-                return 0.0;
-            }
-
-            var taikoCurrent = (TaikoDifficultyHitObject)current;
-
-            double objectStrain = 0.0;
-
-            if (previousHitType != null && taikoCurrent.HitType != previousHitType)
-            {
-                // The colour has changed.
-                objectStrain = 1.0;
-
-                if (monoHistory.Count < 2)
-                {
-                    // There needs to be at least two streaks to determine a strain.
-                    objectStrain = 0.0;
-                }
-                else if ((monoHistory[^1] + currentMonoLength) % 2 == 0)
-                {
-                    // The last streak in the history is guaranteed to be a different type to the current streak.
-                    // If the total number of notes in the two streaks is even, nullify this object's strain.
-                    objectStrain = 0.0;
-                }
-
-                objectStrain *= repetitionPenalties();
-                currentMonoLength = 1;
-            }
-            else
-            {
-                currentMonoLength += 1;
-            }
-
-            previousHitType = taikoCurrent.HitType;
-            return objectStrain;
+            double difficulty = ((TaikoDifficultyHitObject)current).Colour?.EvaluatedDifficulty ?? 0;
```

Comment:
I'm really unsure about this, ideally we wouldn't have to store object difficulties in the `DifficultyHitObject` itself (which is why evaluators exist), but the use case (judging by `ColourEvaluator`) is sorta complex. I don't think it necessarily requires pre-evaluation though.

If my understanding of the underlying flow and system is correct, then the first solution that comes to mind for me would be something like (briefly): additionally encode a marker for whether an object is the first object in one of the groupings, and check for it in `StrainValueOf`, if yes then call the evaluator difficulty method which'd calculate and add the relevant colour difficulty.

Might be easier to elaborate on Discord if that's confusing or if there's something I missed that requires this sort of setup.

Reply:
Index properties sound much better to me than storing difficulty when preprocessing

Reply:
Hmm yeah, I will try to implement this (not sure how much complexity, if any, will this add to the encoding process)

Reply:
Implemented [here](https://github.com/vunyunt/osu/tree/stateless-colour-evaluator), two potential issues are:
 - Encoding classes are now tightly coupled (not sure how much of an issue this actually is), they now store both their parent and children (within `Payload`). This is required due to each encoding's difficulty taking into account its parent's difficulty.
 - Due to the same reason, there are some unnecessary computation, since each time a child encoding's difficulty is calculated its parent's difficulty need to be calculated too. Though I don't expect this to be that significant.
 
 A solution I can think of (aside from just accepting the above or just leaving it pre-evaluated) is to precompute encoding difficulty and only have the evaluator add them together. This does solve the issue of storing difficulty within `TaikoDifficultyHitObject(Colour)`, but the evaluator is now just an addition, and it's mostly the same as pre-evaluation.

Reply:
Applied in e4086b058bb3052a89c41c709bd6483b3fe15809